File: 2020/05/advent202005.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strings"
)

func main() {
	data, err := os.ReadFile("input202005.txt")
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	seatIDs := []int{}
	for _, line := range input {
		seatIDs = append(seatIDs, calcSeatID((line)))
	}
	sort.Ints(seatIDs)
	fmt.Println(part1(seatIDs))
	fmt.Println(part2(seatIDs))
}

func calcSeatID(line string) int {
	row, col := 0.0, 0.0
	for i := 0; i < 7; i++ {
		if line[i] == "B"[0] {
			row += math.Pow(2, float64(6-i))
		}
	}
	for i := 7; i < 10; i++ {
		if line[i] == "R"[0] {
			col += math.Pow(2, float64(9-i))
		}
	}
	return int(8*row + col)
}

func part1(seatIDs []int) int {
	return seatIDs[len(seatIDs)-1]
}

func part2(seatIDs []int) int {
	for i, n := range seatIDs {
		if seatIDs[i+1] == n+2 {
			return n + 1
		}
	}
	return -1
}

File: 2020/05/advent202005.py
from typing import List, Tuple

class Seat:
  def __init__(self, seat: str) -> None:
    self.seat = seat
    self.row, self.col = self._calculate_seat()
    self.seatID = self.row * 8 + self.col

  def _calculate_seat(self) -> Tuple[int, int]:
    row = int(self.seat[:7].replace('F', '0').replace('B', '1'), 2)
    col = int(self.seat[7:].replace('L', '0').replace('R', '1'), 2)
    return row, col

def part1(seatIDs: List[int]) -> int:
  return seatIDs[-1]

def part2(seatIDs: List[int]) -> int:
  for i in range(len(seatIDs) - 1):
    if seatIDs[i] == seatIDs[i + 1] - 2:
      return seatIDs[i] + 1
  return -1

if __name__ == "__main__":
  with open('input.txt','r') as file:
    lines = [line.strip() for line in file]

  seats: List[Seat] = [Seat(line) for line in lines]
  seatIDs: List[int] = [seat.seatID for seat in seats]
  seatIDs.sort()
  print(part1(seatIDs))
  print(part2(seatIDs))

